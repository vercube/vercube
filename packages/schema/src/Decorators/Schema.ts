import { BaseDecorator, createDecorator, Inject } from '@vercube/di';
import type { DeepPartial, MetadataTypes } from '@vercube/core';
import type { RouteConfig } from '@asteasolutions/zod-to-openapi';
import { SchemaRegistry } from '../Services/SchemaRegistry';
import { SchemaBodyResolver } from '../Resolvers/SchemaBodyResolver';
import { SchemaQueryParamsResolver } from '../Resolvers/SchemaQueryParamsResolver';

// oxlint-disable-next-line no-empty-object-type
interface SchemaDecoratorOptions extends Omit<RouteConfig, 'method' | 'path'> {}

/**
 * A decorator class for handling OpenAPI schema registration for routes.
 *
 * This class extends BaseDecorator and is used to register route schemas
 * with the SchemaRegistry. It also applies schema resolvers for responses,
 * body, and query parameters.
 *
 * @extends {BaseDecorator<SchemaDecoratorOptions>}
 */
class SchemaDecorator extends BaseDecorator<SchemaDecoratorOptions> {
  @Inject(SchemaRegistry)
  private readonly gSchemaRegistry: SchemaRegistry;

  public override async created(): Promise<void> {
    // move to the end of the event loop
    await new Promise((resolve) => setTimeout(resolve, 0));

    // get method metadata object
    const _methodMeta = this.prototype.__metadata.__methods[
      this.propertyName
    ] as MetadataTypes.Method;

    const _schema: DeepPartial<RouteConfig> = {
      method: _methodMeta.method!.toLowerCase() as any,
      path: _methodMeta.url!,
      ...(this.options as any),
    };

    // Use resolvers
    SchemaBodyResolver(_methodMeta, _schema);
    SchemaQueryParamsResolver(_methodMeta, _schema);

    // register schema
    this.gSchemaRegistry.register(_schema as RouteConfig);
  }
}

/**
 * A decorator function for registering OpenAPI schema metadata for a route.
 *
 * This function creates an instance of the SchemaDecorator class and registers
 * the schema with the specified options.
 *
 * @param {SchemaDecoratorOptions} params - The schema options for the route.
 * @returns {Function} - The decorator function.
 */
export function Schema(params: SchemaDecoratorOptions): Function {
  return createDecorator(SchemaDecorator, params);
}
