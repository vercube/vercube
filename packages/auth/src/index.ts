// Decorators
export * from './Decorators/Authenticate';
export * from './Decorators/Authorize';

// Middlewares
export * from './Middleware/AuthenticationMiddleware';
export * from './Middleware/AuthorizationMiddleware';

// Providers
export * from './Services/AuthenticationProvider';
export * from './Services/AuthorizationProvider';
