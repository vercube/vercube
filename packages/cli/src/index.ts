import { defineCommand, runMain } from 'citty';
import { version } from '../package.json';
import { buildCommand } from './commands/build';
import { devCommand } from './commands/dev';
import { initCommand } from './commands/init';

const main = defineCommand({
  meta: {
    name: 'Vercube',
    version,
    description: 'Vercube CLI',
  },
  subCommands: {
    build: buildCommand,
    dev: devCommand,
    init: initCommand,
  },
});

runMain(main);