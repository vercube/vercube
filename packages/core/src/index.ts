// Common modules
export * from './Common/App';
export * from './Common/CreateApp';

// Config
export * from './Config/Config';
export * from './Config/Loader';

// Http decorators
export * from './Decorators/Http/Body';
export * from './Decorators/Http/Connect';
export * from './Decorators/Http/Controller';
export * from './Decorators/Http/Delete';
export * from './Decorators/Http/Get';
export * from './Decorators/Http/Head';
export * from './Decorators/Http/Header';
export * from './Decorators/Http/Headers';
export * from './Decorators/Http/Options';
export * from './Decorators/Http/Param';
export * from './Decorators/Http/Patch';
export * from './Decorators/Http/Post';
export * from './Decorators/Http/Put';
export * from './Decorators/Http/QueryParam';
export * from './Decorators/Http/QueryParams';
export * from './Decorators/Http/Request';
export * from './Decorators/Http/Response';
export * from './Decorators/Http/Trace';
export * from './Decorators/Http/SetHeader';
export * from './Decorators/Http/Status';
export * from './Decorators/Http/Redirect';
export * from './Decorators/Http/Middleware';
export * from './Decorators/Http/MultipartFormData';

// Hooks
export * from './Decorators/Hooks/Listen';
export * from './Services/Hooks/HooksService';

// Plugins
export * from './Services/Plugins/BasePlugin';

// Middlewares
export * from './Services/Middleware/BaseMiddleware';

// Errors
export * from './Errors/HttpError';
export * from './Errors/Http/BadRequestError';
export * from './Errors/Http/ForbiddenError';
export * from './Errors/Http/InternalServerError';
export * from './Errors/Http/MethodNotAllowedError';
export * from './Errors/Http/NotAcceptableError';
export * from './Errors/Http/NotFoundError';
export * from './Errors/Http/UnauthorizedError';

// Types
export * from './Types/CommonTypes';
export * from './Types/HooksTypes';
export * from './Types/HttpTypes';
export * from './Types/MetadataTypes';
export * from './Types/ConfigTypes';

// Utils
export * from './Utils/Utils';
