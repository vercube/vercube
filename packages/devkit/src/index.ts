/** Common */
export * from './Common/App';

/** Build */
export * from './Build/Build';
export * from './Build/Watch';

/** Dev server */
export * from './Server/DevServer';
