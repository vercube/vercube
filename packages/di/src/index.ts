// Common
export * from './Common/BaseDecorators';

// Decorators
export * from './Decorators/Inject';
export * from './Decorators/InjectOptional';
export * from './Decorators/Init';
export * from './Decorators/Destroy';
// Domain
export * from './Domain/Container';

// Types
export * from './Types/IOCTypes';

// Utils
export * from './Utils/Utils';