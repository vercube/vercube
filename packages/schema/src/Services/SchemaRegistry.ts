import { OpenAPIRegistry, OpenApiGeneratorV3, type RouteConfig } from '@asteasolutions/zod-to-openapi';

/**
 * Manages the OpenAPI schema registry and provides utilities to generate OpenAPI components
 * from registered Zod schemas using the @asteasolutions/zod-to-openapi library.
 *
 * This class is intended to be used for collecting and generating OpenAPI-compatible
 * schema definitions for API documentation and validation purposes.
 */
export class SchemaRegistry {
  /**
   * Internal registry instance for storing OpenAPI schema definitions.
   * @private
   */
  private fRegistry: OpenAPIRegistry = new OpenAPIRegistry();

  /**
   * Registers a route configuration with the registry.
   * @param cfg - The route configuration to register.
   */
  public register(cfg: RouteConfig): void {
    this.fRegistry.registerPath(cfg);
  }

  /**
   * Generates OpenAPI components from the registered schemas.
   *
   * @async
   * @returns {Promise<unknown>} A promise that resolves to the generated OpenAPI components object.
   */
  public async generateSchema(): Promise<ReturnType<OpenApiGeneratorV3['generateDocument']>> {
    return new OpenApiGeneratorV3(this.fRegistry.definitions).generateDocument({
      openapi: '3.0.0',
      info: {
        title: 'API',
        version: '1.0.0',
      },
    });
  }
}
