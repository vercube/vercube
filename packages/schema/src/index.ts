import { extendZodWithOpenApi } from '@asteasolutions/zod-to-openapi';
import { z } from 'zod';

// Decorators
export * from './Decorators/Schema';

// Plugin
export * from './Plugins/SchemaPlugin';

// extends ZOD schema
extendZodWithOpenApi(z);

export { z };
