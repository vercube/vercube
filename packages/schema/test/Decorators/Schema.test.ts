import { describe, it, expect, beforeEach } from 'vitest';
import { type App, createApp } from '@vercube/core';
import { SchemaPlugin } from '../../src';
import { MockController } from '../Utils/Schema.mock';
import { SchemaRegistry } from '../../src/Services/SchemaRegistry';

describe('SchemaDecorator', () => {
  let app: App;

  beforeEach(async () => {
    app = await createApp({
      setup: async (app) => {
        app.addPlugin(SchemaPlugin);
      },
    });

    app.container.expand((container) => {
      container.bind(MockController);
    });

    await new Promise((resolve) => setTimeout(resolve, 0));
  });

  it('Should register schema correctly', async () => {
    const registry = app.container.get(SchemaRegistry);
    const schema = await registry.generateSchema();

    const path = schema.paths['/mock/'];

    expect(path).toBeDefined();
    expect(path.get).toBeDefined();
    expect(path?.get?.responses['200']).toBeDefined();
    expect(
      path?.get?.responses['200']?.content['application/json']?.schema,
    ).toBeDefined();
    expect(
      path?.get?.responses['200']?.content['application/json']?.schema
        ?.properties.message,
    ).toBeDefined();
    expect(
      path?.get?.responses['200']?.content['application/json']?.schema
        ?.properties.message.type,
    ).toBe('string');
  });

  it('should generate schema correctly', async () => {
    const registry = app.container.get(SchemaRegistry);
    const schema = await registry.generateSchema();

    expect(schema).toMatchSnapshot();
  });
});
