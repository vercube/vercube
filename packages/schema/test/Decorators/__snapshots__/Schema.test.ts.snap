// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SchemaDecorator > should generate schema correctly 1`] = `
{
  "components": {
    "parameters": {},
    "schemas": {},
  },
  "info": {
    "title": "API",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/mock/mock/": {
      "get": {
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "message": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "message",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Mock schema",
          },
        },
      },
    },
  },
}
`;
